rust/user-info-fetcher/src/util.rs

```diff
@@ -1,8 +1,54 @@
-use reqwest::RequestBuilder;
+use hyper::StatusCode;
+use reqwest::{RequestBuilder, Response};
 use serde::de::DeserializeOwned;
+use snafu::{ResultExt, Snafu};
 
-pub async fn send_json_request<T: DeserializeOwned>(
-    req: RequestBuilder,
-) -> Result<T, reqwest::Error> {
-    req.send().await?.error_for_status()?.json().await
+#[derive(Snafu, Debug)]
+pub enum Error {
+    #[snafu(display("failed to execute request"))]
+    HttpRequest { source: reqwest::Error },
+
+    #[snafu(display("failed to parse json response"))]
+    ParseJson { source: reqwest::Error },
+
+    #[snafu(display("response was an HTTP error: {text}"))]
+    HttpErrorResponse { status: StatusCode, text: String },
+
+    #[snafu(display("response was an HTTP error with undecodable text"))]
+    HttpErrorResponseUndecodableText {
+        status: StatusCode,
+        encoding_error: reqwest::Error,
+    },
+}
+
+pub async fn send_json_request<T: DeserializeOwned>(req: RequestBuilder) -> Result<T, Error> {
+    // make the request
+    let response = req.send().await.context(HttpRequestSnafu)?;
+    // check for client or server errors
+    let non_error_response = get_non_error_response(response).await?;
+    // parse the result
+    let result = non_error_response.json().await.context(ParseJsonSnafu)?;
+    Ok(result)
+}
+
+/// takes a Response and checks whether it is an error. If so, parse the reqwest Error
+/// and create our own error type with more context added. We do this because the plain
+/// reqwest error does not give any response body context.
+async fn get_non_error_response(response: Response) -> Result<Response, Error> {
+    let status = response.status();
+    // good response
+    if status.is_success() || status.is_informational() || status.is_redirection() {
+        Ok(response)
+    }
+    // error response branch -> get the text and raise an error
+    else {
+        match response.text().await {
+            Ok(text) => HttpErrorResponseSnafu { status, text }.fail(),
+            Err(encoding_error) => HttpErrorResponseUndecodableTextSnafu {
+                status,
+                encoding_error,
+            }
+            .fail(),
+        }
+    }
```

Comment:
**thought** (non-blocker): I would probably have flipped the logic to keep it closer to how reqwest does it in `error_for_status()`.

I would also avoid the `else` branch, as per:
- [Else considered smelly](https://wiki.c2.com/?ElseConsideredSmelly=)
- [Return early pattern](https://medium.com/swlh/return-early-pattern-3d18a41bba8)

```rust
    let status = response.status();

    // http client/server error response -> get the text and raise an error
    if status.is_client_error() || is_server_error {
        match response.text().await {
            Ok(text) => return HttpErrorResponseSnafu { status, text }.fail(),
            Err(encoding_error) => return HttpErrorResponseUndecodableTextSnafu {
                status,
                encoding_error,
            }
            .fail(),
        }
    }

    Ok(response)
```

Reply:
_I think instead of the explicit returns, a `?` can be used after the `match`._

Reply:
I fully agree on using the protective / guard pattern as much as possible. Returning as early as possible is always nice to have!

Reply:
I flipped the logic but couldn't get rid of the `else` when using `?` as well. I decided I'd rather have `?` and `else` than `return`s and no else: https://github.com/stackabletech/opa-operator/pull/510/commits/8d4427b495ed21fc2edd89cc05460a2207f2bb2b

Reply:
It should work. The `?` shouldn't be after fail() I don't thing, but rather after the match. Because the `match` will always return a Result::Err (because of what fail is doing in both arms):

```rust
    let status = response.status();
    if status.is_client_error() || status.is_server_error() {
        let url = response.url().to_owned();
        match response.text().await {
            Ok(text) => HttpErrorResponseSnafu { status, url, text }.fail(),
            Err(encoding_error) => HttpErrorResponseUndecodableTextSnafu {
                status,
                url,
                encoding_error,
            }
            .fail(),
        }?
    }

    Ok(response)
```

Maybe I'm missing something

Reply:
Oh, a use after move seems to occur, but I think this works (prepending the match with an early return):

```rust
async fn error_for_status(response: Response) -> Result<Response, Error> {
    let status = response.status();
    if status.is_client_error() || status.is_server_error() {
        let url = response.url().to_owned();
        return match response.text().await {
            Ok(text) => HttpErrorResponseSnafu { status, url, text }.fail(),
            Err(encoding_error) => HttpErrorResponseUndecodableTextSnafu {
                status,
                url,
                encoding_error,
            }
            .fail(),
        };
    }

    Ok(response)
}
```

Reply:
https://github.com/stackabletech/opa-operator/pull/510/commits/587db50beb6b5d15a943b6989c2490697706577e

Alright, fine with using a return instead of the question mark